### PR TITLE
Feat/abolfazl kabiri

### PR DESCRIPTION
**Pull request by KabiriAbolfazl**

### CODE REVIEW

**Comment by thisisnabi** [comment]
That was so hard to understand, Keep it simple, stupid, please follow our example.
And I will present your implementation in next coder session.

**Comment by KabiriAbolfazl** [comment]
> That was so hard to understand, Keep it simple, stupid, please follow our example. And I will present your implementation in next coder session.

I agree with you.

**Comment by KabiriAbolfazl** [comment]
> > That was so hard to understand, Keep it simple, stupid, please follow our example. And I will present your implementation in next coder session.
> 
> I agree with you.

I'll push a new implementation

File: src/Abolfazl Kabiri/PipelineDesignPattern.SimpleImplement/Framework/IranianAuthenticationStep.cs
Comment (mohammadKarimi) [inline src/Abolfazl Kabiri/PipelineDesignPattern.SimpleImplement/Framework/IranianAuthenticationStep.cs:1]: - File Scoped namespace
- invert IF
- Naming Convention ( EndPoint vs EndPoind??)
- Remove line space
-

**Review by mohammadKarimi — CHANGES_REQUESTED** [review]
totally your coding style is perfect, but you have to aware of needless complexity. 
Please Keep it Simple.

File: src/Abolfazl Kabiri/PipelineDesignPattern.SimpleImplement/Framework/IranianAuthenticationStep.cs
Comment (KabiriAbolfazl) [reply src/Abolfazl Kabiri/PipelineDesignPattern.SimpleImplement/Framework/IranianAuthenticationStep.cs:1]: Thanks for great review and your hints dear Mohammad

File: src/Abolfazl Kabiri/PipelineDesignPattern.SimpleImplement/Framework/AmericanAuthenticationStep.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+{
+    public Func<T> Func { get; set; }
+    public void Exceute(IPipelineContext context)
+    {
+        "Starting auth".Dump();
+        ValidateIpAddress(context);
+        Func();
+        "End auth".Dump();
+    }
+    private bool ValidateIpAddress(IPipelineContext context)
```
Comment (mohammadKarimi) [inline src/Abolfazl Kabiri/PipelineDesignPattern.SimpleImplement/Framework/AmericanAuthenticationStep.cs:17]: In both IranianAuthenticationStep and AmericanAuthenticationStep, the ValidateIpAddress method throws an InvalidCastException if the IP address cannot be parsed. Consider using a more specific exception type

File: src/Abolfazl Kabiri/PipelineDesignPattern.SimpleImplement/Framework/AmericanAuthenticationStep.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    {
```
Comment (mohammadKarimi) [inline src/Abolfazl Kabiri/PipelineDesignPattern.SimpleImplement/Framework/AmericanAuthenticationStep.cs:18]: The IranianAuthenticationStep and AmericanAuthenticationStep classes share similar code.
DRY 🥇

File: src/Abolfazl Kabiri/PipelineDesignPattern.SimpleImplement/Framework/IranianAuthenticationStep.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    }
+    private bool ValidateIpAddress(IPipelineContext context)
+    {
+        IPAddress parsedIpAddress;
+        IPAddress.TryParse(context.RequestIpAddress, out parsedIpAddress);
+        if (!parsedIpAddress.ToString().Equals(context.RequestIpAddress))
+            throw new InvalidCastException("Ip address is not valid");
+        else
+        {
+            if (!context.RequestIpAddress.Split('.')[0].Equals("194")) // sample for american ip range
```
Comment (mohammadKarimi) [inline src/Abolfazl Kabiri/PipelineDesignPattern.SimpleImplement/Framework/IranianAuthenticationStep.cs:25]: magic string. for practice it is ok, consider don't use like this in real application.

File: src/Abolfazl Kabiri/PipelineDesignPattern.SimpleImplement/Pipeline/Pipeline.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    IPipelineStep _startPoint;
+    public Pipeline(IPipelineContext context)
+    {
+        _context = context;
+    }
+    public void SetStartProccessPoint(IPipelineStep startPointStep)
+    {
+        _startPoint = startPointStep;
+    }
+    public void ExecutePipeline()
```
Comment (mohammadKarimi) [inline src/Abolfazl Kabiri/PipelineDesignPattern.SimpleImplement/Pipeline/Pipeline.cs:14]: for single line of method, 
you can use => instead { }

**Review by mohammadKarimi — CHANGES_REQUESTED** [review]

**Review by mohammadKarimi — APPROVED** [review]
Perfect.

**Pull request merged**